src/fomo.jl

```diff
@@ -0,0 +1,399 @@
+  solver = FomoSolver(nlp)
+  solver_specific = Dict(:avgβmax => T(0.))
+  stats = GenericExecutionStats(nlp;solver_specific=solver_specific)
+  return solve!(solver, nlp, stats; kwargs...)
+end
+
+function SolverCore.reset!(solver::FomoSolver{T}) where {T}
+  fill!(solver.m,0)
+  solver
+end
+
+SolverCore.reset!(solver::FomoSolver, ::AbstractNLPModel) = reset!(solver)
+
+mutable struct FoSolver{T, V} <: AbstractFirstOrderSolver
+  x::V
+  g::V
+  c::V
+  α::T
+end
+
+function FoSolver(nlp::AbstractNLPModel{T, V}) where {T, V}
+  x = similar(nlp.meta.x0)
+  g = similar(nlp.meta.x0)
+  c = similar(nlp.meta.x0)
+  return FoSolver{T, V}(x, g, c, T(0))
+end
+
+@doc (@doc FomoSolver) function R2(nlp::AbstractNLPModel{T, V}; kwargs...) where {T, V}
+  solver = FoSolver(nlp)
+  stats = GenericExecutionStats(nlp)
+  if haskey(kwargs,:σmin)
+    return solve!(solver, nlp, stats; step_backend = r2_step(), αmax = 1/kwargs[:σmin], kwargs...)
+  else
+    return solve!(solver, nlp, stats; step_backend = r2_step(), kwargs...)
+  end
+end
+
+function SolverCore.reset!(solver::FoSolver{T}) where {T}
+  solver
+end
+
+SolverCore.reset!(solver::FoSolver, ::AbstractNLPModel) = reset!(solver)
+
+function SolverCore.solve!(
+  solver::AbstractFirstOrderSolver,
+  nlp::AbstractNLPModel{T, V},
+  stats::GenericExecutionStats{T, V};
+  callback = (args...) -> nothing,
+  x::V = nlp.meta.x0,
+  atol::T = √eps(T),
+  rtol::T = √eps(T),
+  η1::T = T(eps(T)^(1 / 4)),
+  η2::T = T(0.95),
+  γ1::T = T(1/2),
+  γ2::T = T(2),
+  γ3::T = T(1/2),
+  αmax::T = 1/eps(T),
+  max_time::Float64 = 30.0,
+  max_eval::Int = -1,
+  max_iter::Int = typemax(Int),
+  β::T = T(0.9),
+  θ1::T = T(0.1),
+  θ2::T = T(eps(T)^(1/3)),
+  verbose::Int = 0,
+  step_backend = r2_step(),
+  σmin = nothing # keep consistency with R2 interface. kwargs immutable, can't delete it in `R2`
+) where {T, V}
+  use_momentum = typeof(solver) <: FomoSolver
+  unconstrained(nlp) || error("fomo should only be called on unconstrained problems.")
+  
+  reset!(stats)
+  start_time = time()
+  set_time!(stats, 0.0)
+
+  x = solver.x .= x
+  ∇fk = solver.g
+  c = solver.c
+  momentum = use_momentum ? solver.m : nothing # not used if no momentum
+  d = use_momentum ? solver.d : solver.g # g = d if no momentum
+  p = use_momentum ? solver.p : nothing # not used if no momentum
+  set_iter!(stats, 0)
+  set_objective!(stats, obj(nlp, x))
+
+  
+  grad!(nlp, x, ∇fk)
+  norm_∇fk = norm(∇fk)
+  set_dual_residual!(stats, norm_∇fk)
+
+  solver.α = init_alpha(norm_∇fk,step_backend)
+  
+  # Stopping criterion: 
+  ϵ = atol + rtol * norm_∇fk
+  optimal = norm_∇fk ≤ ϵ
+  if optimal
+    @info("Optimal point found at initial point")
+    if !use_momentum
+      @info @sprintf "%5s  %9s  %7s  %7s " "iter" "f" "‖∇f‖" "σ"
+      @info @sprintf "%5d  %9.2e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk 1/solver.α
+    else
+      @info @sprintf "%5s  %9s  %7s  %7s " "iter" "f" "‖∇f‖" "α"
```

Comment:
The header seems to be the same in both cases.

Reply:
Also, are you sure the test is on whether or not momentum is used? Or on whether R2 or TR is used? Same comment elsewhere in the code.

Reply:
Yes, use_momentum variable (first line of `solve!`) is set from `solver` type. The R2 or TR step are given by the `step_backend` kwarg, which is given only to the `init_alpha` and `step_mult` functions which are the only "step-sensitive" part of the code.

Reply:
What I’m saying is that in one case, you print α, and in the other you print 1/α. That’s related to R2 vs TR, not to momentum.

Reply:
Exact, there's a mistake here.

Reply:
Use σ and Δ in the header, but print α and 1/α.